File: mobile-app/src/components/GForceOverlay.js
import React, { useMemo } from 'react';
import { View, Text, StyleSheet } from 'react-native';

const G = 9.81; // m/s² per G

// ── Force thresholds (in G) ──
// Below DEAD_ZONE the edge glow is invisible; only engine vibration and
// road noise should fall in this range.
const DEAD_ZONE_G = 0.05;
// Yellow/amber ceiling — forces above this start transitioning to red.
const YELLOW_CEIL_G = 0.20;
// Fully red at or above this value.
const RED_FLOOR_G = 0.40;
// Used to cap the opacity ramp so it doesn't exceed ~0.8.
const MAX_DISPLAY_G = 0.80;

// ── Glow geometry ──
const GLOW_LAYERS = 5;
const LAYER_PX = 12; // thickness of each gradient band

const LAYER_INDICES = Array.from({ length: GLOW_LAYERS }, (_, i) => i);

// ── Colour helpers ──

// Returns [r, g, b] interpolated from amber → red based on G-force.
function forceRGB(g) {
  if (g <= YELLOW_CEIL_G) return [245, 158, 11]; // amber  #F59E0B
  if (g >= RED_FLOOR_G) return [239, 68, 68];     // red    #EF4444
  const t = (g - YELLOW_CEIL_G) / (RED_FLOOR_G - YELLOW_CEIL_G);
  return [
    Math.round(245 + (239 - 245) * t),
    Math.round(158 + (68 - 158) * t),
    Math.round(11 + (68 - 11) * t),
  ];
}

// Base opacity for the innermost (brightest) glow layer.
function baseOpacity(g) {
  if (g <= DEAD_ZONE_G) return 0;
  const t = Math.min((g - DEAD_ZONE_G) / (MAX_DISPLAY_G - DEAD_ZONE_G), 1);
  return 0.2 + t * 0.6; // 0.2 → 0.8
}

/**
 * Full-screen overlay that lights up the edges of the screen based on
 * real-time accelerometer data.
 *
 * • Cornering  → left or right edge glows (yellow → red).
 * • Braking    → top edge glows (yellow → red).
 *
 * Uses `pointerEvents="none"` so touches pass through to the UI below.
 *
 * @param {Object}  props.acceleration  {x, y, z} in m/s² (gravity removed)
 * @param {boolean} props.isActive      only render when a ride is in progress
 */
export default function GForceOverlay({ acceleration, isActive }) {
  const { lateralG, brakingG, turnDir } = useMemo(() => {
    if (!isActive || !acceleration) {
      return { lateralG: 0, brakingG: 0, turnDir: null };
    }

    // Lateral: phone x-axis = left/right in both flat and mounted orientations.
    const latG = Math.abs(acceleration.x) / G;

    // Longitudinal: orientation-agnostic — take the larger of |y| and |z|.
    // After gravity removal both are ≈0 at rest; whichever spikes during
    // braking is the axis aligned with the road.
    const lonG = Math.max(
      Math.abs(acceleration.y),
      Math.abs(acceleration.z),
    ) / G;

    // Centrifugal force direction:
    //   +x → pushed right → car is turning LEFT  → light up LEFT edge
    //   −x → pushed left  → car is turning RIGHT → light up RIGHT edge
    const dir = latG > DEAD_ZONE_G
      ? (acceleration.x > 0 ? 'left' : 'right')
      : null;

    return { lateralG: latG, brakingG: lonG, turnDir: dir };
  }, [acceleration?.x, acceleration?.y, acceleration?.z, isActive]);

  if (!isActive) return null;

  const latOp = baseOpacity(lateralG);
  const brkOp = baseOpacity(brakingG);

  const [lr, lg, lb] = forceRGB(lateralG);
  const [br, bg, bb] = forceRGB(brakingG);

  const peakG = Math.max(lateralG, brakingG);
  const [pr, pg, pb] = peakG > 0 && peakG === lateralG
    ? forceRGB(lateralG)
    : forceRGB(brakingG);

  // Badge colour: use force colour when active, muted when idle
  const badgeColor = peakG > DEAD_ZONE_G
    ? `rgb(${pr},${pg},${pb})`
    : '#94A3B8';

  return (
    <View style={StyleSheet.absoluteFill} pointerEvents="none">

      {/* ── Left edge glow (turning left) ── */}
      {turnDir === 'left' && LAYER_INDICES.map(i => (
        <View
          key={`l${i}`}
          style={{
            position: 'absolute',
            top: 0,
            bottom: 0,
            left: i * LAYER_PX,
            width: LAYER_PX,
            backgroundColor: `rgba(${lr},${lg},${lb},${(
              latOp * (1 - i / GLOW_LAYERS)
            ).toFixed(3)})`,
          }}
        />
      ))}

      {/* ── Right edge glow (turning right) ── */}
      {turnDir === 'right' && LAYER_INDICES.map(i => (
        <View
          key={`r${i}`}
          style={{
            position: 'absolute',
            top: 0,
            bottom: 0,
            right: i * LAYER_PX,
            width: LAYER_PX,
            backgroundColor: `rgba(${lr},${lg},${lb},${(
              latOp * (1 - i / GLOW_LAYERS)
            ).toFixed(3)})`,
          }}
        />
      ))}

      {/* ── Top edge glow (braking / longitudinal force) ── */}
      {brkOp > 0 && LAYER_INDICES.map(i => (
        <View
          key={`t${i}`}
          style={{
            position: 'absolute',
            left: 0,
            right: 0,
            top: i * LAYER_PX,
            height: LAYER_PX,
            backgroundColor: `rgba(${br},${bg},${bb},${(
              brkOp * (1 - i / GLOW_LAYERS)
            ).toFixed(3)})`,
          }}
        />
      ))}

      {/* ── G-force badge (always visible during ride) ── */}
      <View style={styles.badge}>
        <Text style={[styles.badgeLabel, { color: badgeColor }]}>
          {peakG.toFixed(1)}
        </Text>
        <Text style={styles.badgeUnit}>G</Text>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  badge: {
    position: 'absolute',
    bottom: 104,
    alignSelf: 'center',
    flexDirection: 'row',
    alignItems: 'baseline',
    backgroundColor: 'rgba(0, 0, 0, 0.55)',
    paddingHorizontal: 14,
    paddingVertical: 5,
    borderRadius: 14,
  },
  badgeLabel: {
    fontSize: 22,
    fontWeight: '900',
    letterSpacing: 0.5,
  },
  badgeUnit: {
    fontSize: 13,
    fontWeight: '800',
    color: '#94A3B8',
    marginLeft: 2,
  },
});

File: mobile-app-kmp/shared/src/commonMain/kotlin/com/roadready/ui/components/GForceOverlay.kt
package com.roadready.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import com.roadready.ui.theme.*
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

private fun formatGForce(value: Float): String {
    val rounded = (value * 10f).roundToInt() / 10f
    val parts = rounded.toString().split(".")
    return "${parts[0]}.${(parts.getOrElse(1) { "0" }).take(1).padEnd(1, '0')}"
}

private const val G = 9.81f

// Force thresholds (in G)
private const val DEAD_ZONE_G = 0.10f
private const val YELLOW_CEIL_G = 0.25f
private const val RED_FLOOR_G = 0.40f
private const val MAX_DISPLAY_G = 0.80f

private const val GLOW_LAYERS = 5
private const val LAYER_FRACTION = 0.025f // each layer = 2.5% of dimension

/**
 * RGB color interpolated from amber → red as G-force increases.
 */
private fun forceColor(g: Float): Color {
    if (g <= YELLOW_CEIL_G) return Color(0xFFF59E0B) // amber
    if (g >= RED_FLOOR_G) return Color(0xFFEF4444)   // red
    val t = (g - YELLOW_CEIL_G) / (RED_FLOOR_G - YELLOW_CEIL_G)
    val r = (245f + (239f - 245f) * t).roundToInt().coerceIn(0, 255)
    val gVal = (158f + (68f - 158f) * t).roundToInt().coerceIn(0, 255)
    val b = (11f + (68f - 11f) * t).roundToInt().coerceIn(0, 255)
    return Color(r, gVal, b)
}

/**
 * Base opacity for the innermost (brightest) glow layer.
 */
private fun baseOpacity(g: Float): Float {
    if (g <= DEAD_ZONE_G) return 0f
    val t = min((g - DEAD_ZONE_G) / (MAX_DISPLAY_G - DEAD_ZONE_G), 1f)
    return 0.2f + t * 0.6f
}

private enum class TurnDirection { LEFT, RIGHT }

/**
 * Full-screen overlay that lights up edges based on real-time accelerometer data.
 *
 * - Cornering → left or right edge glows (amber → red).
 * - Braking / longitudinal force → top edge glows (amber → red).
 * - Center displays the total G-force magnitude.
 *
 * The overlay does not consume touch events; layer it above content that should
 * remain interactive.
 */
@Composable
fun GForceOverlay(
    accelerationX: Float,
    accelerationY: Float,
    accelerationZ: Float,
    modifier: Modifier = Modifier,
) {
    val lateralG = abs(accelerationX) / G
    val brakingG = max(abs(accelerationY), abs(accelerationZ)) / G
    val totalG = sqrt(
        (accelerationX * accelerationX +
                accelerationY * accelerationY +
                accelerationZ * accelerationZ)
    ) / G

    val turnDir: TurnDirection? = remember(accelerationX, lateralG) {
        when {
            lateralG <= DEAD_ZONE_G -> null
            accelerationX > 0 -> TurnDirection.LEFT
            else -> TurnDirection.RIGHT
        }
    }

    val latOpacity = baseOpacity(lateralG)
    val brkOpacity = baseOpacity(brakingG)
    val latColor = forceColor(lateralG)
    val brkColor = forceColor(brakingG)

    Box(modifier = modifier.fillMaxSize()) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            val w = size.width
            val h = size.height
            val layerW = w * LAYER_FRACTION
            val layerH = h * LAYER_FRACTION

            // Left edge glow (turning left)
            if (turnDir == TurnDirection.LEFT) {
                for (i in 0 until GLOW_LAYERS) {
                    val alpha = latOpacity * (1f - i.toFloat() / GLOW_LAYERS)
                    drawRect(
                        color = latColor.copy(alpha = alpha),
                        topLeft = Offset(i * layerW, 0f),
                        size = Size(layerW, h),
                    )
                }
            }

            // Right edge glow (turning right)
            if (turnDir == TurnDirection.RIGHT) {
                for (i in 0 until GLOW_LAYERS) {
                    val alpha = latOpacity * (1f - i.toFloat() / GLOW_LAYERS)
                    drawRect(
                        color = latColor.copy(alpha = alpha),
                        topLeft = Offset(w - (i + 1) * layerW, 0f),
                        size = Size(layerW, h),
                    )
                }
            }

            // Top edge glow (braking / longitudinal)
            if (brkOpacity > 0f) {
                for (i in 0 until GLOW_LAYERS) {
                    val alpha = brkOpacity * (1f - i.toFloat() / GLOW_LAYERS)
                    drawRect(
                        color = brkColor.copy(alpha = alpha),
                        topLeft = Offset(0f, i * layerH),
                        size = Size(w, layerH),
                    )
                }
            }
        }

        // G-force magnitude in center
        if (totalG > DEAD_ZONE_G) {
            val magColor = forceColor(totalG)
            Column(
                modifier = Modifier.align(Alignment.Center),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Text(
                    text = formatGForce(totalG),
                    fontSize = 28.sp,
                    fontWeight = FontWeight.Black,
                    color = magColor,
                )
                Text(
                    text = "G",
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Bold,
                    color = magColor.copy(alpha = 0.7f),
                )
            }
        }
    }
}
